### PR TITLE
Remove most `assert()` calls (issue 8506) 

### LINKED ISSUE #8506: Remove error() and re-inspect assert() calls.
There is lots of confusion how error() is used. It also may mislead people when reading source code -- error() hides `throw new Error()`, so it's not easy to see control flow for human and JS analyzers cannot complain about dead code.

The same issue can be found with assert(). Normally assert() is used to define precondition for the following code, means that code shall never have its check `false`. (In other languages it's usually defined as hard stop for the program). By hiding checks and throw inside assert(), we also potentially removing JIT ability to make some assumptions.

- [x] Introduce InvalidPDFError() to raise when PDF is malformed instead of generic Error()
- [x] We need to review code and change `error(_)` => `throw new Error(_)`
- [ ] We need to review code and change `assert(_,?)` => `if (!_) throw new Error(_)` for non-assert-like calls (test for it is: it's impossible to remove this check/assert without changing logic of the program and affect the testing)

### PR DESCRIPTION
**Pull request by Snuffleupagus**
This replaces `assert` calls with `throw new FormatError()`/`throw new Error()`.
In a few places, throwing an `Error` (which is what `assert` meant) isn't correct since the enclosing function is supposed to return a `Promise`, hence some cases were changed to `Promise.reject(...)` and similarily for `createPromiseCapability` instances.

Fixes #8506.

@yurydelendik Based on https://github.com/mozilla/pdf.js/issues/8506#issuecomment-307455596 in particular: I wasn't really sure if we should leave some, and if so *which* ones, of the current `assert` calls in place.
~~This patch is currently replacing *all* of the them, but I figured that might not be what you actually want here which is why I've marked the PR as `work-in-progress` and would very much appreciate feedback on this!~~

### CODE REVIEW

File: src/core/evaluator.js [hunk 1]
```diff
@@ -765,7 +767,9 @@ var PartialEvaluator = (function PartialEvaluatorClosure() {
         }
         this.fontCache.put('id_' + fontID, fontCapability.promise);
       }
-      assert(fontID, 'The "fontID" must be defined.');
+      if (!fontID) {
```
Comment (yurydelendik) [inline src/core/evaluator.js]: looks like fontID generated by internal logic. if we can tell that out logic produces non-empty id, we can keep assert.

File: src/core/image.js [hunk 2]
```diff
@@ -525,7 +523,9 @@ var PDFImage = (function PDFImageClosure() {
           }
           if (this.needsDecode) {
             // Invert the buffer (which must be grayscale if we reached here).
-            assert(kind === ImageKind.GRAYSCALE_1BPP);
```
Comment (yurydelendik) [inline src/core/image.js]: looks like a good candidate for assert, if needsDecode == true is driven only by internal logic (and not use data)

File: src/core/network.js [hunk 3]
```diff
@@ -291,7 +291,9 @@ PDFNetworkStream.prototype = {
   },
 
   getFullReader: function PDFNetworkStream_getFullReader() {
-    assert(!this._fullRequestReader);
```
Comment (yurydelendik) [inline src/core/network.js]: internal logic, let's keep it

File: src/core/parser.js [hunk 4]
```diff
@@ -157,7 +157,9 @@ var Parser = (function ParserClosure() {
         } else if (state === 1) {
           state = (ch === I) ? 2 : 0;
         } else {
-          assert(state === 2);
```
Comment (yurydelendik) [inline src/core/parser.js]: internal logic, let's keep it

File: src/core/pattern.js [hunk 5]
```diff
@@ -403,7 +405,9 @@ Shadings.Mesh = (function MeshClosure() {
 
   function buildFigureFromPatch(mesh, index) {
     var figure = mesh.figures[index];
-    assert(figure.type === 'patch', 'Unexpected patch mesh figure');
```
Comment (yurydelendik) [inline src/core/pattern.js]: mostly internal logic, I think we can keep it

File: src/core/worker.js [hunk 6]
```diff
@@ -232,7 +232,9 @@ var PDFWorkerStream = (function PDFWorkerStreamClosure() {
            rangeReader._enqueue(args.chunk);
            return true;
          });
-         assert(found);
```
Comment (yurydelendik) [inline src/core/worker.js]: internal logic, let's keep it

File: src/core/worker.js [hunk 7]
```diff
@@ -254,7 +256,9 @@ var PDFWorkerStream = (function PDFWorkerStreamClosure() {
     },
 
     getFullReader: function PDFWorkerStream_getFullReader() {
-      assert(!this._fullRequestReader);
```
Comment (yurydelendik) [inline src/core/worker.js]: internal logic, let's keep it

File: src/display/font_loader.js [hunk 8]
```diff
@@ -155,7 +155,9 @@ if (typeof PDFJSDev === 'undefined' || !PDFJSDev.test('MOZCENTRAL')) {
   FontLoader.prototype.queueLoadingCallback =
       function FontLoader_queueLoadingCallback(callback) {
     function LoadLoader_completeRequest() {
-      assert(!request.end, 'completeRequest() cannot be called twice');
```
Comment (yurydelendik) [inline src/display/font_loader.js]: internal logic, let's keep it

File: src/shared/util.js [hunk 9]
```diff
@@ -1634,7 +1635,6 @@ export {
   FormatError,
   arrayByteLength,
   arraysToBytes,
-  assert,
```
Comment (yurydelendik) [inline src/shared/util.js]: The asserts in this file looks like internal logic, let's keep them... and export

**Review by yurydelendik — COMMENTED** [review]
Here is my suggestions for acceptable asserts.

**Comment by Snuffleupagus** [comment]
@yurydelendik Thanks a lot for your feedback, I've revert all of the `assert`s that you suggested.

/botio test

**Review by timvandermeij — APPROVED** [review]

**Comment by timvandermeij** [comment]
Nice work, thanks!

**Pull request merged**